apps/system/js/audio_channel_manager.js

```diff
@@ -0,0 +1,370 @@
+    },
+
+    /**
+     * Handle the audio channel.
+     *
+     * @param {Object} policy The policy to handle the audio channel.
+     * @param {AudioChannelController} policy.audioChannel The audio channel.
+     * @param {Boolean} policy.isAllowedToPlay Play or pause it.
+     * @param {Boolean} policy.isNeededToFadeOut Fade out or not.
+     * @param {Boolean} policy.isNeededToVibrate Vibrate or not.
+     * @param {Boolean} policy.isResumedWhenOtherEnds
+     * Resume when other audio channel ends if as true.
+     * @return {Promise}
+     */
+    _handleAudioChannel: function(policy) {
+      var promise;
+      var audioChannel = policy.audioChannel;
+      if (policy.isAllowedToPlay) {
+        promise =
+          this._playAudioChannel(audioChannel, policy.isNeededToFadeOut);
+      } else {
+        promise =
+          this._interruptAudioChannel(
+            audioChannel, policy.isResumedWhenOtherEnds
+          );
+      }
+      if (policy.isNeededToVibrate && this._isVibrateEnabled) {
+        audioChannel.vibrate();
+      }
+      return promise;
+    },
+
+    /**
+     * Play the audio channel and
+     * manage the state of `_activeAudioChannels` map.
+     *
+     * @param {AudioChannelController} audioChannel The audio channel.
+     * @param {Boolean} isFadedOut Fade out or not.
+     */
+    _playAudioChannel: function(audioChannel, isFadedOut) {
+      var promise = isFadedOut ?
+        audioChannel.fadeOut() :
+        audioChannel.play();
+      promise.then(() => {
+        return new Promise((resolve) => {
+          this._activeAudioChannels.set(
+            audioChannel.instanceID,
+            audioChannel
+          );
+          resolve();
+          this.debug('Playing ' + audioChannel.channelName);
+          this.debug('In fade out mode: ' + isFadedOut);
+        });
+      });
+      return promise;
+    },
+
+    /**
+     * Interrupt the active audio channel and
+     * put it into the `_interruptedAudioChannels` map.
+     * If it would like to be resumed when other audio channel ends,
+     * put it into `_audioChannelsResumeWhenOtherEnds` array.
+     *
+     * @param {AudioChannelController} audioChannel The audio channel.
+     * @param {Boolean} isResumedWhenOtherEnds
+     * Resume when other audio channel ends if as true.
+     */
+    _interruptAudioChannel: function(audioChannel, isResumedWhenOtherEnds) {
+      return audioChannel.pause().then(() => {
+        return new Promise((resolve, reject) => {
+          var id = audioChannel.instanceID;
+          this._activeAudioChannels.delete(id);
+          var backgroundResumeIndex;
+          if (isResumedWhenOtherEnds) {
+            var length =
+              this._audioChannelsResumeWhenOtherEnds.push(audioChannel);
+            backgroundResumeIndex = length - 1;
+          }
+          this._interruptedAudioChannels.set(id, {
+            audioChannel: audioChannel,
+            backgroundResumeIndex: backgroundResumeIndex
```

Comment:
Read the patch multiple times and then I really don't think you need to keep this variable...